base/tuple.jl

```diff
@@ -20,7 +20,10 @@ endof(t::Tuple) = length(t)
 size(t::Tuple, d) = d==1 ? length(t) : throw(ArgumentError("invalid tuple dimension $d"))
 getindex(t::Tuple, i::Int) = getfield(t, i)
 getindex(t::Tuple, i::Real) = getfield(t, convert(Int, i))
-getindex(t::Tuple, r::AbstractArray{<:Any,1}) = tuple([t[ri] for ri in r]...)
+function getindex(t::Tuple, r::AbstractArray{<:Any,1})
+    s = Ref(start(r))
+    ntuple(j -> ((i, s[]) = next(r, s[]); t[i]), length(r))
+end
```

Comment:
Turns out that inference works different when you directly return something like `c1 ? r1 : c2 ? r2 : c3 ? r3 ...` to when you assign that to a variable and then return the variable. So basically what improved things was assigning the tuple before returning it.

I won't leave this version as is as it messes inference when using this function inside macros in some circumstances (because of the closure). But I have a slightly different version that doesn't have that problem.